File: core/src/main/kotlin/eldoria/core/world/LevelProgression.kt
package eldoria.core.world

import eldoria.core.model.DiceFormula
import eldoria.core.model.DieType
import eldoria.core.model.PlayerCharacter
import eldoria.core.model.StatBlock
import kotlin.random.Random

/**
 * Character level (1..50) is its own track, separate from skill mastery
 * (see SkillProgression): defeating enemies and completing quests grants
 * experience points; enough banked xp levels the character up. Higher
 * world-map difficulty tiers (1..5) are tougher fights that pay out more
 * xp, so hunting in a harder biome zone is the "grind faster" lever.
 *
 * A level-up bumps every basic stat slightly -- health and strength gain
 * the most, agility/willpower a little -- and each level demands more xp
 * than the last, same escalating-difficulty shape used everywhere else in
 * the engine.
 */
object LevelProgression {
    const val MAX_CHARACTER_LEVEL = 50

    /** Banked xp required to advance from `level` to `level + 1`. */
    fun xpToNextLevel(level: Int): Int = level * level * 20 + 100

    /** Xp a kill of the given world difficulty tier (1..5) pays out. Harder tier, bigger reward. */
    fun xpForDefeating(tier: Int, rng: Random): Int {
        val t = tier.coerceIn(1, 5)
        return DiceFormula(t, DieType.D20, t * 10).roll(rng) * t
    }

    /** Apply a batch of xp (from a kill, a quest reward, whatever), leveling up as many times as it covers. */
    fun applyExperience(player: PlayerCharacter, xpGained: Int, rng: Random): PlayerCharacter {
        var level = player.level
        var xp = player.experience + xpGained
        var strength = player.strength
        var agility = player.agility
        var willpower = player.willpower
        var maxHealth = player.maxHealth
        var currentHealth = player.currentHealth
        var maxStamina = player.maxStamina
        var currentStamina = player.currentStamina
        var pendingPerkChoices = player.pendingPerkChoices

        while (level < MAX_CHARACTER_LEVEL && xp >= xpToNextLevel(level)) {
            xp -= xpToNextLevel(level)
            level++

            val healthGain = DiceFormula(2, DieType.D8, 3).roll(rng)
            maxHealth += healthGain
            currentHealth += healthGain
            val staminaGain = DiceFormula(1, DieType.D8, 2).roll(rng)
            maxStamina += staminaGain
            currentStamina += staminaGain
            if (DiceFormula(1, DieType.D20).roll(rng) <= 12) strength++ // ~60% chance
            if (DiceFormula(1, DieType.D20).roll(rng) <= 8) agility++ // ~40% chance
            if (DiceFormula(1, DieType.D20).roll(rng) <= 8) willpower++ // ~40% chance
            if (level % 5 == 0) pendingPerkChoices++
        }
        if (level >= MAX_CHARACTER_LEVEL) {
            level = MAX_CHARACTER_LEVEL
            xp = 0
        }

        val strMod = StatBlock.modifierOf(strength)
        val agiMod = StatBlock.modifierOf(agility)
        val armorClass = 10 + agiMod + level / 5
        val attackBonus = level / 4 + strMod
        val speed = (20 + agiMod * 2 + level / 10).coerceAtLeast(5)
        val unarmedDamage = DiceFormula(1 + level / 15, if (level < 25) DieType.D6 else DieType.D8, strMod)

        return player.copy(
            level = level,
            experience = xp,
            strength = strength,
            agility = agility,
            willpower = willpower,
            maxHealth = maxHealth,
            currentHealth = currentHealth,
            maxStamina = maxStamina,
            currentStamina = currentStamina,
            armorClass = armorClass,
            speed = speed,
            attackBonus = attackBonus,
            unarmedDamage = unarmedDamage,
            pendingPerkChoices = pendingPerkChoices,
        )
    }
}

File: eldoria/world/level_progression.py
"""Character level (1..50): its own track, separate from skill mastery.

Defeating enemies and completing quests grants experience; enough banked xp
levels the character up. A level-up bumps every basic stat slightly, and
each level demands more xp than the last.
"""
from __future__ import annotations

import random
from dataclasses import replace

from eldoria.models import DiceFormula, DieType, PlayerCharacter, StatBlock

MAX_CHARACTER_LEVEL = 50


def xp_to_next_level(level: int) -> int:
    """Banked xp required to advance from `level` to `level + 1`."""
    return level * level * 20 + 100


def xp_for_defeating(tier: int, rng: random.Random) -> int:
    """Xp a kill of the given world difficulty tier (1..5) pays out. Harder tier, bigger reward."""
    t = max(1, min(5, tier))
    return DiceFormula(t, DieType.D20, t * 10).roll(rng) * t


def apply_experience(player: PlayerCharacter, xp_gained: int, rng: random.Random) -> PlayerCharacter:
    """Apply a batch of xp (from a kill, a quest reward, whatever), leveling up as many times as it covers."""
    level = player.level
    xp = player.experience + xp_gained
    strength = player.strength
    agility = player.agility
    willpower = player.willpower
    max_health = player.max_health
    current_health = player.current_health
    max_stamina = player.max_stamina
    current_stamina = player.current_stamina
    pending_perk_choices = player.pending_perk_choices

    while level < MAX_CHARACTER_LEVEL and xp >= xp_to_next_level(level):
        xp -= xp_to_next_level(level)
        level += 1

        health_gain = DiceFormula(2, DieType.D8, 3).roll(rng)
        max_health += health_gain
        current_health += health_gain
        stamina_gain = DiceFormula(1, DieType.D8, 2).roll(rng)
        max_stamina += stamina_gain
        current_stamina += stamina_gain
        if DiceFormula(1, DieType.D20).roll(rng) <= 12:  # ~60% chance
            strength += 1
        if DiceFormula(1, DieType.D20).roll(rng) <= 8:  # ~40% chance
            agility += 1
        if DiceFormula(1, DieType.D20).roll(rng) <= 8:  # ~40% chance
            willpower += 1
        if level % 5 == 0:
            pending_perk_choices += 1

    if level >= MAX_CHARACTER_LEVEL:
        level = MAX_CHARACTER_LEVEL
        xp = 0

    str_mod = StatBlock.modifier_of(strength)
    agi_mod = StatBlock.modifier_of(agility)
    armor_class = 10 + agi_mod + level // 5
    attack_bonus = level // 4 + str_mod
    speed = max(5, 20 + agi_mod * 2 + level // 10)
    unarmed_damage = DiceFormula(1 + level // 15, DieType.D6 if level < 25 else DieType.D8, str_mod)

    return replace(
        player,
        level=level,
        experience=xp,
        strength=strength,
        agility=agility,
        willpower=willpower,
        max_health=max_health,
        current_health=current_health,
        max_stamina=max_stamina,
        current_stamina=current_stamina,
        armor_class=armor_class,
        speed=speed,
        attack_bonus=attack_bonus,
        unarmed_damage=unarmed_damage,
        pending_perk_choices=pending_perk_choices,
    )
